decider-bcl2fastq/src/main/java/ca/on/oicr/pde/deciders/utils/BarcodeAndBasesMask.java

```diff
@@ -0,0 +1,178 @@
+
+        if (barcode.getBarcodeOne().length() > 0) {
+            basesMaskBuilder.setIndexOneIncludeLength(barcode.getBarcodeOne().length());
+        }
+
+        if (barcode.getBarcodeTwo() != null && barcode.getBarcodeTwo().length() > 0) {
+            basesMaskBuilder.setIndexTwoIncludeLength(barcode.getBarcodeTwo().length());
+            basesMaskBuilder.setIndexTwoIgnoreLength(Integer.MAX_VALUE);
+        }
+
+        return basesMaskBuilder.createBasesMask();
+    }
+
+    public static BasesMask calculateBasesMask(Barcode barcode, BasesMask runBasesMask) throws DataMismatchException {
+        Barcode sequencedBarcode = tryApplyingBasesMask(barcode, runBasesMask);
+        BasesMask barcodeBasesMask = calculateBasesMask(sequencedBarcode);
+
+        BasesMask.BasesMaskBuilder basesMaskBuilder = new BasesMask.BasesMaskBuilder();
+
+        //read one
+        basesMaskBuilder.setReadOneIncludeLength(runBasesMask.getReadOneIncludeLength());
+        basesMaskBuilder.setReadOneIgnoreLength(runBasesMask.getReadOneIgnoreLength());
+
+        //index one
+        if (runBasesMask.getIndexOneIncludeLength() == null && barcodeBasesMask.getIndexOneIncludeLength() == null) {
+            throw new DataMismatchException("Unexpected state, runBasesMask = [" + runBasesMask.toString() + "],"
+                    + " barcodeBasesMask = [" + barcodeBasesMask.toString() + "]");
+        } else if (runBasesMask.getIndexOneIncludeLength() == null && barcodeBasesMask.getIndexOneIncludeLength() != null) {
+            basesMaskBuilder.setIndexOneIncludeLength(barcodeBasesMask.getIndexOneIncludeLength());
+            basesMaskBuilder.setIndexOneIgnoreLength(barcodeBasesMask.getIndexOneIgnoreLength());
+        } else if (runBasesMask.getIndexOneIncludeLength() != null && barcodeBasesMask.getIndexOneIncludeLength() == null) {
+            throw new DataMismatchException("Unexpected state, runBasesMask = [" + runBasesMask.toString() + "],"
+                    + " barcodeBasesMask = [" + barcodeBasesMask.toString() + "]");
+        } else if (runBasesMask.getIndexOneIncludeLength() != null && barcodeBasesMask.getIndexOneIncludeLength() != null) {
+            basesMaskBuilder.setIndexOneIncludeLength(barcodeBasesMask.getIndexOneIncludeLength());
+            basesMaskBuilder.setIndexOneIgnoreLength(barcodeBasesMask.getIndexOneIgnoreLength());
+        } else {
+            throw new DataMismatchException("Unexpected state, runBasesMask = [" + runBasesMask.toString() + "],"
+                    + " barcodeBasesMask = [" + barcodeBasesMask.toString() + "]");
+        }
+
+        //index two
+        if (runBasesMask.getIndexTwoIncludeLength() == null && barcodeBasesMask.getIndexTwoIncludeLength() == null) {
+            basesMaskBuilder.setIndexTwoIncludeLength(null);
+            basesMaskBuilder.setIndexTwoIgnoreLength(null);
+        } else if (runBasesMask.getIndexTwoIncludeLength() == null && barcodeBasesMask.getIndexTwoIncludeLength() != null) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This branch's code block is the same as the block for the branch on line 122. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1871)